src/assets.cc

```diff
@@ -0,0 +1,159 @@
+
+
+    // The acual function hooks
+
+    SubHook dslDbAddMembers; // We need this to add the entries, as we need to do it after the DB table is set up
+
+    SubHook dslDbTryOpenDetour;
+    SubHook dslDbTryOpenDetour2;
+
+    void (*dsl_db_add_members)   (lua_state*);
+
+    void* (*dsl_db_try_open)   (void* /* this */, void* name, void* ext, void* target, void* transport, void* z);
+    void* (*dsl_db_try_open_2)   (void* /* this */, void* name, void* ext, void* target, void* transport, void* z);
+
+    int (*dsl_db_do_resolve_1) (dsl::DB *_this, dsl::idstring*, dsl::idstring*, void *abc, void *def);
+    void (*dsl_fss_open) (dsl::Archive *output, dsl::FileSystemStack **_this, std::string const*);
+
+    static void*
+    dt_dsl_db_try_open(void* _this, void* name, void* ext, void* target, void* transport, void* z) {
+        hook_remove(dslDbTryOpenDetour);
+
+        //printf("Hello, World: %ld %lx.%lx %ld\n", *((uint64_t*) name), *((uint64_t*) target), *((uint64_t*) ext), sizeof(std::string));
+
+        return dsl_db_try_open(_this, name, ext, target, transport, z);
+    }
+
+    static void*
+    dt_dsl_db_try_open_2(dsl::Archive *target, dsl::DB* db, dsl::idstring* ext, dsl::idstring* name, void* misc_object, dsl::Transport* transport) {
+        hook_remove(dslDbTryOpenDetour2);
+
+        hash_t hash(*name, *ext);
+        if(custom_assets.count(hash)) {
+            string str = custom_assets[hash];
+            dsl_fss_open(target, &db->stack, &str);
+            return target;
+        }
+
+        return dsl_db_try_open_2(target, db, ext, name, misc_object, transport);
+
+        /* // Code for doing the same thing as the original function (minus mod_override support):
+        int result = dsl_db_do_resolve_1(
+            db,
+            ext,
+            name,
+            misc_object,
+            (void*) ((char*)db->ptr4 + 40)
+        );
+
+        if ( result < 0 )
+        {
+            // Couldn't find the asset
+            *target = dsl::Archive("", 0LL, 0LL, 0LL, 1, 0LL);
+        }
+        else
+        {
+            char *ptr = (char*) db;
+    #define add_and_dereference(value) ptr = *(char**)(ptr + value)
+            add_and_dereference(80);
+            add_and_dereference(56);
+    #undef add_and_dereference
+            unsigned int val = *(unsigned int*) (ptr + 24 + result * 32);
+            transport->vt->f_open(target, transport, val);
+        }
+        return target;
+        */
+    }
+
+    static void dt_dsl_db_add_members(lua_state *L)
+    {
+        hook_remove(dslDbAddMembers);
+
+        dsl_db_add_members(L);
+
+        lapi::assets::setup(L);
+    }
+
+    void init_asset_hook(void *dlHandle)
+    {
+#define setcall(symbol,ptr) *(void**) (&ptr) = dlsym(dlHandle, #symbol); 
+        setcall(_ZN3dsl6MainDB11add_membersEP9lua_State, dsl_db_add_members);
+
+        setcall(_ZN3dsl2DB8try_openINS_16LanguageResolverEEENS_7ArchiveENS_8idstringES4_RKT_RKNS_9TransportE, dsl_db_try_open);
```

Comment:
You are relying on a specifig name mangling.. This is neither defined by the standard nor necessarily persistent across compiler versions. 
Would it be possible to demangle the names?

Reply:
While it would be nice not to, I'm not sure the how I'd go about doing so - short of reading all the names in the binary, demangling them and checking them.

Also, it's worth noting that we're doing the same thing in `hook.cc`, and AFAIK those signatues haven't broken in the lifetime of BLT4L. Observations with the VR and non-VR versions of the windows binaries suggest they're very unlikely to change their compiler setup.

Reply:
@hellow554 we need to rely on mangling in order to maintain compatibility with PAYDAY, unfortunately. DieselX is built with Clang on libcxx. If we want to use as many demangled names as possible, we should expect users to use the exact same development environment for gameplay.

That being said, GCC and Clang behave (for the most part) identically when mangling names, so you can reliably build a matching class/namespace structure and have it mangle properly.